### DISCUSSION TITLE
Cannot send file

### DISCUSSION THREAD

**Discussion by shintaii** [saleweaver/python-amazon-sp-api#122] [Q&A]
I have created a .tsv file (flat file) When uploading that file, manually on sellercentral.amazon.com everything works out fine and the file is processed. 

When sending the same file using `submit_feed()` it errors with:
```
Feed Processing Summary:
	Number of records processed		0
	Number of records successful		0

original-record-number	sku	error-code	error-type	error-message
0		90011	Fatal	The file's column headings are missing or invalid. To correct this error, download the template again and use that new copy, or insert the correct column headings as row two of your existing file.
```

File used:
`sku	product-id	product-id-type	price	minimum-seller-allowed-price	maximum-seller-allowed-price	item-condition	quantity	add-delete	will-ship-internationally	expedited-shipping	standard-plus	item-note	fulfillment-center-id
0889698521550	889698521550	4	40.95	40.95	40.95	11	0	a	4	N	Y		`

File is tab separated and utf8 encoded.

**Comment by saleweaver** [comment]
I can’t reproduce the error without knowing your code. 
On 11 Apr 2021, 16:04 +0200, Patrick Gerrits ***@***.***>, wrote:


I have created a .tsv file (flat file) When uploading that file, manually on sellercentral.amazon.com everything works out fine and the file is processed.

When sending the same file using submit_feed() it errors with:


Feed Processing Summary:
        Number of records processed             0
        Number of records successful            0

original-record-number  sku     error-code      error-type      error-message
0               90011   Fatal   The file's column headings are missing or invalid. To correct this error, download the template again and use that new copy, or insert the correct column headings as row two of your existing file.


File used:
sku product-id product-id-type price minimum-seller-allowed-price maximum-seller-allowed-price item-condition quantity add-delete will-ship-internationally expedited-shipping standard-plus item-note fulfillment-center-id 0889698521550 889698521550 4 40.95 40.95 40.95 11 0 a 4 N Y

File is tab separated and utf8 encoded.

**Reply by shintaii** [reply]
````
def submit_feed(self, feed_type, content_type='text/tsv'):
        if feed_type == 'POST_FLAT_FILE_INVLOADER_DATA':
            file = self.store.create_amazon_inventory_loader_flat_file()

         r = AmazonApi.Feeds(
            marketplace=AmazonBase.Marketplaces.NL,
            refresh_token=self.spapi_oauth_code
        ).submit_feed(
            feed_type,
            file,
            content_type
        )

        return r
````

````
    def create_amazon_inventory_loader_flat_file(self):
        from inventory.models import AmazonFlatFile
        import csv
        from io import StringIO
        from django.core.files.base import ContentFile

        products = self.product_set.all()[:10]

        csv_buffer = StringIO()
        csv_writer = csv.writer(csv_buffer, delimiter='\t')
        csv_writer.writerow(['sku', 'product-id', 'product-id-type', 'price', 'minimum-seller-allowed-price',
                             'maximum-seller-allowed-price', 'item-condition', 'quantity', 'add-delete',
                             'will-ship-internationally', 'expedited-shipping', 'standard-plus', 'item-note', 'fulfillment-center-id'])

        for p in products:
            csv_writer.writerow([
                p.sku,  # sku
                p.ean,  # product_id
                4,  # product-id-type
                p.price_marketplace,  # price
                p.price_marketplace,  # minimum-seller-allowed-price
                p.price_marketplace,  # maximum-seller-allowed-price
                11,  # item-condition
                0,  # quantity
                'a',  # add-delete
                4,  # will-ship-internationally
                'N',  # expedited-shipping
                'Y',  # standard-plus
                None,  # item-note
                None,  # fulfillment-center-id
            ])

       csv_file = ContentFile(csv_buffer.getvalue().encode('utf8'))

        amz_file = AmazonFlatFile()
        amz_file.flat_file.save('amazon_tsv_import.txt', csv_file)
        amz_file.type='POST_FLAT_FILE_INVLOADER_DATA'
        amz_file.save()

        return amz_file.flat_file
````

```
from marketplaces.models import Amazon
az = Amazon.objects.get(id=2)
r = az.submit_feed('POST_FLAT_FILE_INVLOADER_DATA')
```

**Comment by shintaii** [comment]
Same result:

```
csv_buffer = StringIO()
csv_writer = csv.writer(csv_buffer, delimiter='\t')
csv_writer.writerow(['sku', 'product-id', 'product-id-type', 'price', 'minimum-seller-allowed-price',
                             'maximum-seller-allowed-price', 'item-condition', 'quantity', 'add-delete',
                             'will-ship-internationally', 'expedited-shipping'])

for p in products:
    csv_writer.writerow([
        p.sku,  # sku
        p.ean,  # product_id
        4,  # product-id-type
        p.price_marketplace,  # price
        p.price_marketplace,  # minimum-seller-allowed-price
        p.price_marketplace,  # maximum-seller-allowed-price
        11,  # item-condition
        0,  # quantity
        'a',  # add-delete
        4,  # will-ship-internationally
        None  # expedited-shipping
    ])

 
buf = BytesIO()
buf.write(csv_buffer.getvalue().encode('UTF-8'))
buf.seek(0)
buf.name='test-upload.txt'

r = AmazonApi.Feeds(marketplace=AmazonBase.Marketplaces.NL,refresh_token=az.spapi_oauth_code).submit_feed('POST_FLAT_FILE_INVLOADER_DATA', buf)
```

And also Same result:

```
feed_header = 'sku\tproduct-id\tproduct-id-type\tprice\titem-condition\tquantity\tadd-delete\n'
feed_row_template = '{}\t{}\t{}\t{}\t{}\t{}\t{}'

s = Store.objects.get(id=4)
products = s.product_set.all()[:10]

feed_row_list = []
for product in products:
        feed_row = feed_row_template.format(p.sku, p.ean, p.price_marketplace,4, 11, 0, 'a')
        feed_row_list.append(feed_row)

feed_data = feed_header + '\n'.join(feed_row_list)
feed_data = feed_data.encode('utf8')

buf2 = BytesIO()
buf2.write(feed_data)
buf2.seek(0)
buf2.name='test-upload2.txt'

r = AmazonApi.Feeds(marketplace=AmazonBase.Marketplaces.NL,refresh_token=az.spapi_oauth_code).submit_feed('POST_FLAT_FILE_INVLOADER_DATA', buf2)
```

**Comment by Dennis694** [comment]
Hello,

unfortunately I have the same issue and couldn't figure out how to solve this myself. 
Did you solve this problem @shintaii? Or is there anyone that can provide a working example how to create and submit a feed?

I tried to use the inventory loader as well as the listings loader flat file (see below code for one example product, which I wanted to use for testing). Unfortunately both result in the same error, shown in the initial post. 

`csv_buffer = StringIO()
csv_writer = csv.writer(csv_buffer, delimiter='\t')

csv_writer.writerow(['TemplateType=Offer','Version=2020.000'])
csv_writer.writerow(['sku', 'price', 'quantity', 'product-id', 'product-id-type', 'condition-type', 
                     'condition-note', 'ASIN-hint', 'title', 'product-tax-code', 'operation-type', 'sale-price', 
                     'sale-start-date', 'sale-end-date', 'leadtime-to-ship', 'launch-date', 'is-giftwrap-available', 
                     'is-gift-message-available', 'fulfillment-center-id', 'batteries_required', 'are_batteries_included', 
                     'battery_cell_composition', 'battery_type', 'number_of_batteries', 'battery_weight', 
                     'battery_weight_unit_of_measure', 'number_of_lithium_metal_cells', 
                     'number_of_lithium_ion_cells', 'lithium_battery_packaging', 'lithium_battery_energy_content', 
                     'lithium_battery_energy_content_unit_of_measure', 'lithium_battery_weight', 
                     'lithium_battery_weight_unit_of_measure', 'supplier_declared_dg_hz_regulation1', 
                     'supplier_declared_dg_hz_regulation2', 'supplier_declared_dg_hz_regulation3', 
                     'supplier_declared_dg_hz_regulation4', 'supplier_declared_dg_hz_regulation5', 
                     'hazmat_united_nations_regulatory_id', 'safety_data_sheet_url', 'item_weight', 
                     'item_weight_unit_of_measure', 'item_volume', 'item_volume_unit_of_measure', 
                     'flash_point', 'ghs_classification_class1', 'ghs_classification_class2', 
                     'ghs_classification_class3', 'list_price', 'uvp_list_price'])

csv_writer.writerow([
8516993,  # sku
999.99,
0,
1234567890123,
'EAN',
'New',
None,
None,
None,
None,
None,
'Delete'
])

buf = BytesIO()
buf.write(csv_buffer.getvalue().encode('UTF-8'))
buf.seek(0)
buf.name='test-upload.txt'

res = Feeds(credentials=credentials, marketplace=Marketplaces.DE).submit_feed('POST_FLAT_FILE_LISTINGS_DATA',buf)`

Thanks in advance!

**Reply by shintaii** [reply]
Yes:

when using it as a StringIO it works!

**Reply by Dennis694** [reply]
Thanks, now it is working!

**Reply by GuyTuval** [reply]
Hey @shintaii @McDennis1 !
I am facing the same problem and from what I'm reading you both managed to overcome it.
Would you mind sharing code sample or elaborate where we should use `StringIO`?
Thanks!

**Reply by Dennis694** [reply]
Hi,

sure! I ended up doing the following:
1. Create dataframe, which contains exactly the same columns defined in "columnnames_listingloader" (not included in code snippet below).
2. Create StringIO Buffer in required format and required header rows.
3. Write dataframe values to Buffer.
4. Upload.

`columnnames_listingloader=['sku', 'price', 'quantity', 'product-id', 'product-id-type', 'condition-type', 'condition-note', 'ASIN-hint', 'title', 'product-tax-code', 'operation-type', 'sale-price', 'sale-start-date', 'sale-end-date', 'leadtime-to-ship', 'launch-date', 'is-giftwrap-available', 'is-gift-message-available', 'fulfillment-center-id', 'batteries_required', 'are_batteries_included', 'battery_cell_composition', 'battery_type', 'number_of_batteries', 'battery_weight', 'battery_weight_unit_of_measure', 'number_of_lithium_metal_cells', 'number_of_lithium_ion_cells', 'lithium_battery_packaging', 'lithium_battery_energy_content', 'lithium_battery_energy_content_unit_of_measure', 'lithium_battery_weight', 'lithium_battery_weight_unit_of_measure', 'supplier_declared_dg_hz_regulation1', 'supplier_declared_dg_hz_regulation2', 'supplier_declared_dg_hz_regulation3', 'supplier_declared_dg_hz_regulation4', 'supplier_declared_dg_hz_regulation5', 'hazmat_united_nations_regulatory_id', 'safety_data_sheet_url', 'item_weight', 'item_weight_unit_of_measure', 'item_volume', 'item_volume_unit_of_measure', 'flash_point', 'ghs_classification_class1', 'ghs_classification_class2', 'ghs_classification_class3', 'list_price', 'uvp_list_price']

uploadfile_buffer = StringIO()
uploadfile_writer = csv.writer(uploadfile_buffer, delimiter='\t')
uploadfile_writer.writerow(['TemplateType=Offer','Version=2020.000'])
uploadfile_writer.writerow(columnnames_listingloader)
uploadfile_writer.writerow(columnnames_listingloader)
upload_df.to_csv(uploadfile_buffer, sep ='\t', encoding='utf-8', index=False, header=None) #Write dataframe to StringIO Buffer
uploadfile_buffer.seek(0)
uploadfile_buffer.name=timestamp+"_DailyProductUpdate.txt"

#Upload
res = Feeds(credentials=credentials, marketplace=Marketplaces.DE).submit_feed('POST_FLAT_FILE_LISTINGS_DATA',uploadfile_buffer,'text/tsv')`

**Reply by GuyTuval** [reply]
Thanks!

**Comment by GIT-HUB-SAM** [comment]
HI @shintaii @GuyTuval  @McDennis1 @saleweaver 

I am trying to create a feed document and create a feed for POST_FLAT_FILE_PRICEANDQUANTITYONLY_UPDATE_DATA (Price and Quantity update)

```
from sp_api.api import Feeds

cred = 'mycred'
mark = 'Marketplace_ID'

def create_amazon_inventory_loader_flat_file():
    try:
        columnnames_listingloader = ['sku', 'price', 'minimum-seller-allowed-price', 'maximum-seller-allowed-price',
                                     'quantity', 'handling-time', 'fulfillment-channel']
        data = [['sku1', 1840, '', '', '10', '5', ''],
                ['sku2', 5900, '', '', '10', '5', ''],
                ['sku3', 2310, '', '', '10', '5', '']]
        upload_df = pd.DataFrame(columns= columnnames_listingloader)

        uploadfile_buffer = StringIO()
        uploadfile_writer = csv.writer(uploadfile_buffer, delimiter='\t')
        uploadfile_writer.writerow(['TemplateType=PriceQuantityUpdate', 'Version=2022.000'])
        uploadfile_writer.writerow(columnnames_listingloader)
        uploadfile_writer.writerow(['sku1', 1840, '', '', '10', '5', ''])
        uploadfile_writer.writerow(['sku2', 5900, '', '', '10', '5', ''])
        uploadfile_writer.writerow(['sku3', 2310, '', '', '10', '5', ''])
        upload_df.to_csv(uploadfile_buffer, sep='\t', encoding='utf-8', index=False, header=None)
        # Write dataframe to StringIO Buffer
        uploadfile_buffer.seek(0)
        uploadfile_buffer.name = "flat_file.txt"

        return uploadfile_buffer

file = create_amazon_inventory_loader_flat_file()
print(file.read())

feed_doc_res = Feeds(credentials=cred, marketplace=mark).create_feed_document(file=file, content_type='text/tsv')
print(feed_doc_res.payload)

feed_doc_arg = feed_doc_res.payload.get('feedDocumentId')

res = Feeds(credentials=cred, marketplace=mark).create_feed(
feed_type='POST_FLAT_FILE_PRICEANDQUANTITYONLY_UPDATE_DATA', 
input_feed_document_id=feed_doc_arg)

print(res.payload)

feed_id_arg = res.payload.get('feedId')
feed_id_sts = ''

while feed_id_sts != 'DONE':
    feed_res = Feeds(credentials=cred, marketplace=mark).get_feeds(feedTypes=feed_id_arg)
    print(feed_res.payload)
    feed_id_sts = feed_res.payload.get('processingStatus')
    args = feed_res.payload.get('resultFeedDocumentId')
    print('10 sec sleep')
    time.sleep(10)

feed_doc_res = Feeds(credentials=cred, marketplace=mark).get_feed_document(feedDocumentId=args)
print(feed_doc_res)

file.close()
```

I am getting successful responses for create_feed_document, and create_feed. But then when I see the created feed I see the below response ---
```
Feed Processing Summary:
	Number of records processed		0
	Number of records successful		0

original-record-number	sku	error-code	error-type	error-message
0		90011	Fatal	The file's column headings are missing or invalid. To correct this error, download the template again and use that new copy, or insert the correct column headings as row two of your existing file.
```

I see you guys faced a similar issue but for different file submission, can u guys help me with this

Thanks..,
